src/ZkFold/Symbolic/Base/Num.hs

```diff
@@ -0,0 +1,743 @@
+  , KnownNat
+  , Prime
+  , Finite (..)
+  , FiniteChr (..)
+    -- * Numeric combinators
+  , sum
+  , product
+  , even
+  , odd
+  , fromInteger
+  , fromSemiIntegral
+  , knownNat
+  , order
+  , numberOfBits
+  , characteristic
+  , combineN
+  , combineZ
+  , evalMonoN
+  , evalMonoZ
+  ) where
+
+import           Control.Applicative
+import           Control.Category
+import           Data.Bool
+import           Data.Eq
+import           Data.Foldable       hiding (product, sum, toList)
+import           Data.Functor
+import           Data.Kind
+import           Data.Ord
+import           Data.Ratio
+import           Data.Type.Bool
+import           Data.Type.Equality
+import           GHC.Exts            (proxy#)
+import           GHC.TypeLits        (ErrorMessage (..), TypeError)
+import           GHC.TypeNats        hiding (Mod)
+import qualified GHC.TypeNats        as Type
+import           Prelude             (Int, Integer)
+import qualified Prelude
+
+-- Symbolic field extensions [Arithmetic a : a] should include:
+-- PrimeField x => Symbolic x
+-- PrimeField x => Symbolic (i -> x)
+-- PrimeField x => Symbolic (Circuit x i Par1)
+class
+  ( Field a
+  , Comparable a
+  , FiniteChr a
+  , 3 <= Chr a
+  , PrimeField x
+  , Algebra x a
+  , Chr a ~ Order x
+  ) => Symbolic x a | a -> x where
+
+-- Prime fields should only include:
+-- (SemiIntegral int, Prime p) => PrimeField (int `Mod` p)
+-- and newtypes of int `Mod` p.
+-- p = 2 is ruled out to allow/require trichotomy.
+type PrimeField x =
+  ( Modular x
+  , Finite x
+  , Symbolic x x
+  )
+
+class AdditiveMonoid a where
+  infixl 6 +
+  (+) :: a -> a -> a
+  zero :: a
+
+sum :: (Foldable t, AdditiveMonoid a) => t a -> a
+sum = foldl' (+) zero
+
+class AdditiveMonoid a => AdditiveGroup a where
+  negate :: a -> a
+  negate a = zero - a
+  infixl 6 -
+  (-) :: a -> a -> a
+  a - b = a + negate b
+
+class MultiplicativeMonoid a where
+  infixl 7 *
+  (*) :: a -> a -> a
+  one :: a
+
+product :: (Foldable t, MultiplicativeMonoid a) => t a -> a
+product = foldl' (*) one
+
+class MultiplicativeMonoid a => MultiplicativeGroup a where
+  recip :: a -> a
+  recip a = one / a
+  infixl 7 /
+  (/) :: a -> a -> a
+  a / b = a * recip b
+
+-- from @Natural is the unique homomorphism from the free Semiring
+-- from @Integer is the unique homomorphism from the free Ring
+-- from @Rational is the unique homomorphism from the free Field
+--
+-- prop> from . from = from
+-- prop> from @a @a = id
+class From x a where
+  from :: x -> a
+  default from :: x ~ a => x -> a
+  from = id
+
+type Semiring a =
+  ( AdditiveMonoid a
+  , MultiplicativeMonoid a
+  , From Natural a
+  , From a a
+  , Scalar Natural a
+  , Scalar a a
+  , Exponent Natural a
+  )
+
+type Ring a =
+  ( AdditiveGroup a
+  , MultiplicativeMonoid a
+  , From Natural a
+  , From Integer a
+  , From a a
+  , Scalar Natural a
+  , Scalar Integer a
+  , Scalar a a
+  , Exponent Natural a
+  )
+
+type Field a =
```

Comment:
I still don't agree that `Field` is just an `AdditiveGroup` + `MultiplicativeGroup`, the group of invertible elements of a field is a separate object with its own size (in case we're talking about finite objects), so to treat the whole field as a multiplicative group is not correct IMO